File: libs/power-log-parser/src/lib/parsers/card-present-on-game-start-parser.ts
import { CardType, GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { FullEntity, Game, Node } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

// TODO: Oracle
const Oracle = {
	FindCardCreator(
		_gameState: GameState,
		_entity: FullEntity,
		_node: Node,
	): [string, number] | null {
		return null;
	},
};

export class CardPresentOnGameStartParser implements ActionParser {
	readonly ParserName = 'CardPresentOnGameStartParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === FullEntity &&
			(node.Object as FullEntity).GetTag(GameTag.ZONE) === (Zone.PLAY as number) &&
			node.Parent != null &&
			node.Parent.Type === Game
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const fullEntity = node.Object as FullEntity;
		const cardId = fullEntity.CardId;
		if (
			fullEntity.GetTag(GameTag.CARDTYPE) !== (CardType.MINION as number) ||
			fullEntity.GetTag(GameTag.HAS_BEEN_REBORN) === 1
		) {
			return null;
		}
		const controllerId = fullEntity.GetEffectiveController();
		const startingHealth = fullEntity.GetTag(GameTag.HEALTH);
		const creator = Oracle.FindCardCreator(this.GameState, fullEntity, node);
		return [
			GameEventProvider.Create(
				fullEntity.TimeStamp,
				'CARD_ON_BOARD_AT_GAME_START',
				GameEventHelper.CreateProvider(
					'CARD_ON_BOARD_AT_GAME_START',
					cardId,
					controllerId,
					fullEntity.Id,
					this.StateFacade,
					{
						Health: startingHealth,
						CreatorCardId: creator?.[0] ?? null,
						Tags: fullEntity.GetTagsCopy(),
					},
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/CardPresentOnGameStartParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class CardPresentOnGameStartParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public CardPresentOnGameStartParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(FullEntity)
                && (node.Object as FullEntity).GetTag(GameTag.ZONE) == (int)Zone.PLAY
                && node.Parent != null && node.Parent.Type == typeof(Game);
                //&& !ParserState.ReconnectionOngoing;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var fullEntity = node.Object as FullEntity;
            var cardId = fullEntity.CardId;
            if (fullEntity.GetTag(GameTag.CARDTYPE) != (int)CardType.MINION || fullEntity.GetTag(GameTag.HAS_BEEN_REBORN) == 1)
            {
                return null;
            }
            var controllerId = fullEntity.GetEffectiveController();
            var startingHealth = fullEntity.GetTag(GameTag.HEALTH);
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, null, null);
            var creator = Oracle.FindCardCreator(GameState, fullEntity, node);
            return new List<GameEventProvider> { GameEventProvider.Create(
                fullEntity.TimeStamp,
                "CARD_ON_BOARD_AT_GAME_START",
                GameEvent.CreateProvider(
                    "CARD_ON_BOARD_AT_GAME_START",
                    cardId,
                    controllerId,
                    fullEntity.Id,
                    StateFacade,
                    //gameState,
                    new {
                        Health = startingHealth,
                        CreatorCardId = creator?.Item1,
                        Tags = fullEntity.GetTagsCopy(),
                    }),
                true,
                node) };
        }
    }
}
